BUG: PNG IO fails on corrupt file

issakomi opened the pull request:
S. explanation and link to sample file in the post https://github.com/InsightSoftwareConsortium/ITK/issues/2928#issuecomment-991665972

Also this post is related https://github.com/InsightSoftwareConsortium/ITK/pull/2933#issuecomment-991470346

jhlegarreta commented:
:100: To this effort @issakomi.

Cross-referencing comments from PR #2933
https://github.com/InsightSoftwareConsortium/ITK/pull/2933#discussion_r767291731
https://github.com/InsightSoftwareConsortium/ITK/pull/2933#discussion_r767289300
https://github.com/InsightSoftwareConsortium/ITK/pull/2933#discussion_r767291877

so that corrupted test files are added, and the degenerate situations they trigger are tested.

issakomi commented:
> Cross-referencing comments

Not sure what i could answer here, here are no new "warning" trigger points, no changes in debug output, etc.
Just fixed _setjmp_ wrapping and added one jump point in _ReadImageInformation_ to avoid crash with corrupted example file.

I am afraid i have no more time, already spent 4 full days with broken JPEG and PNG IOs.
Consider this as suggestion and feel free to do with the PR whatever you find correct, close, accept, modify. No problem.

cc @Hconk

jhlegarreta commented:
> Just fixed setjmp wrapping and added one jump point in ReadImageInformation to avoid crash with corrupted example file.

> I am afraid i have no more time, already spent 4 full days with bugs in JPEG and PNG IOs.
Consider this as suggestion and feel free to do with the PR whatever you find correct, close, accept, modify. No problem.

@issakomi I understand. Thanks for the effort. Can at least the corrupted, readable and corrupted, unreadable images used to test this be shared?

Hconk commented:
> > Just fixed setjmp wrapping and added one jump point in ReadImageInformation to avoid crash with corrupted example file.
> 
> > I am afraid i have no more time, already spent 4 full days with bugs in JPEG and PNG IOs.
> > Consider this as suggestion and feel free to do with the PR whatever you find correct, close, accept, modify. No problem.
> 
> @issakomi I understand. Thanks for the effort. Can at least the corrupted, readable and corrupted, unreadable images used to test this be shared?
You can get corrupted png image from here. 
https://github.com/InsightSoftwareConsortium/ITK/issues/2928#issuecomment-991665972

jhlegarreta commented:
> You can get corrupted png image from here.
#2928 (comment)

Thanks @Hconk. I will add it to the testing suite as time permits.

dzenanz reviewed (commented):
LGTM. I guess we should wait until test(s) are added before merging.

issakomi commented:
Use `ITK_TRY_EXPECT_EXCEPTION` here, BTW.

issakomi commented:
Suggestion for test, simply drop if not interested.

```cpp
/*=========================================================================
 *
 *  Copyright NumFOCUS
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *         http://www.apache.org/licenses/LICENSE-2.0.txt
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 *
 *=========================================================================*/

#include "itkPNGImageIO.h"
#include "itkImageFileReader.h"
//#include "itkTestingMacros.h"


int
itkPNGImageIODegenerateCasesTest(int argc, char * argv[])
{
  if (argc != 2)
  {
    std::cerr << "Missing parameters." << std::endl;
    std::cerr << "Usage: " << itkNameOfTestExecutableMacro(argv);
    std::cerr << " inputFilename" << std::endl;
    return EXIT_FAILURE;
  }

  constexpr unsigned int Dimension = 2;
  using PixelType = unsigned char;

  using ImageType = itk::Image<PixelType, Dimension>;

  itk::PNGImageIO::Pointer io = itk::PNGImageIO::New();

  itk::ImageFileReader<ImageType>::Pointer reader = itk::ImageFileReader<ImageType>::New();

  reader->SetFileName(argv[1]);
  reader->SetImageIO(io);

  try
  {
    reader->Update();
  }
  catch
  {
    std::cout << "Corrupted file " << argv[1] << " triggered exception" << std::endl;
  }


  std::cout << "Test finished." << std::endl;
  return EXIT_SUCCESS;
}
```
P.S.
Applicable to **JPEG**
I would recommend the same approach for JPEG too (`s/PNG/JPEG/`). Specially JPEG library has different scenarios. With the logic you could add other files for the same test. All we can do with corrupted files is avoid crash, the rest is up to a library, IMHO.

dzenanz commented:
Looks good to me. Why not add it as the second commit in your branch `pngio`? Then we could merge it, without waiting for Jon's follow-up with improvements.

issakomi commented:
> Why not add it as the second commit in your branch pngio?

Of course. The problem is that I have forgotten how to add files.  I did it once (_singebit_ file). I remember it took me long to get it working. BTW, i could update the testing file for JPEG with same logic, so the same test will work with a file that triggers exception, not depend on
`ITK_TRY_EXPECT_NO_EXCEPTION(reader->Update());`, it will also increase coverage, that exception is not covered by current tests. 
BTW, there is also [minimal issue](https://github.com/InsightSoftwareConsortium/ITK/blob/abb1d19328fe318d77ce78c15d2e29e0c0417b69/Modules/IO/JPEG/test/itkJPEGImageIODegenerateCasesTest.cxx#L30) in current JPEG test
`    std::cerr << "Usage: " << itkNameOfTestExecutableMacro(argv);`
prints pointer `argv`, not the name  `argv[0]`

dzenanz commented:
[Data addition](https://github.com/InsightSoftwareConsortium/ITK/blob/master/Documentation/Data.md) document is not long. [Web interface](https://github.com/InsightSoftwareConsortium/ITK/blob/master/Documentation/UploadBinaryData.md#upload-via-the-web-interface) is the most convenient for uploading a small number of files.

dzenanz commented:
There is also a [visual guide](https://itk.org/ITKExamples/Documentation/Contribute/UploadBinaryData.html).

issakomi commented:
I have reused _itkPNGImageIOTest3.cxx_, it is already for corrupted files with `ITK_TRY_EXPECT_EXCEPTION`.

issakomi commented:
BTW, the fix is minimal. There are more question.
In Read function are two setjmp points, one was wrapped (not working), another not wrapped. Probably long ago there were workaround how to shut up warnings, but preserve test. It could be improved.
Do we want PNG print output or not? The empty ITK custom function (to suppress debug from the library) is set at the very end of the story. There is a lot of PNG debug output in tests log. Not clear.

dzenanz reviewed (approved):
Keeping the current situation regarding the PNG error and warning messages is fine. I don't have a preference for making it more consistent by more or less messages.
